### DISCUSSION TITLE
CR2 or DNG orientation issue, and a general EXIF preservation Q

### DISCUSSION THREAD

**Discussion by peterh337** [ImageMagick/ImageMagick#2142] [Help]
Hi All,

We are using Imagemagick (under Centos 8) to convert CR2 to jpeg.

Landscape CR2 files work fine but portrait ones are not being rotated during the conversion. I can put the CR2 up somewhere if that helps; it is from a Canon G7X. The rotation is handled correctly by Canon's Digital Photo Pro v4 file converter (under windows) so it is definitely there.

The other option would be to use the orientation tag in the resulting jpeg to rotate the jpeg as a second step (with a 100% quality setting) but there is no EXIF left in the jpeg post-conversion.

In this project, it would be nice to preserve EXIF data generally. I have read various past posts about EXIF not being preserved and people using separate tools to transplant the EXIF. Has anything changed in this area or is there some better method?

Perhaps another method is to have some code which extracts the CR2 orientation tag directly and then asks Imagemagick to rotate the image accordingly. Is there any code out there for this?

The same issue would arise with DNG which would be nice to support but it seems to be a worse moving target because every fancy camera mfg is "improving" its DNG... Only Pentax AFAIK stick to the standard version.

Many thanks for any input.

**Comment by snibgo** [comment]
Doesn't "-auto-orient" work?

**Comment by peterh337** [comment]
auto-orient is command line parameter, no option in php lib.

We appear to have solved it using this
https://stackoverflow.com/questions/19456036/detect-exif-orientation-and-rotate-image-using-imagemagick

I wonder why Imagemagick strips out EXIF when image editors have not had problems preserving it. Even the ancient ACDSEE v5 (which came on a computer magazine cover around 1999) does it right, and still works today, blindingly fast as you would expect :) I know the social media arguments (anti stalking measures to e.g. drop the GPS data) but this should be user selectable.

**Comment by snibgo** [comment]
IM is an image processor. It ignores exif data that has no impact on the image, so GPS coordinates etc. You can use exiftool to copy exifdata from one file to another.

(Personally, I would like IM to have the power of exiftool to read, edit and write exif data. And IM should have the power of G'MIC's language. I can dream. Meanwhile, we need different tools for different jobs.)

If you use the linked PHP code to read the orientation metadata and rotate the image, be aware that it doesn't set the metadata to "top-left". So software such as viewers may rotate the image again, creating a wrong result. You should either set the correct orientation (see https://phpimagick.com/Imagick/setImageOrientation?imagePath=Lorikeet&orientationType=4 ), or strip all the metadata.

**Comment by peterh337** [comment]
I must be misunderstanding but a jpeg coming out of IM has no EXIF at all, so unless you did something else to insert EXIF in there, an image viewer is not going to rotate it.

Actually there are around three ways to rotate a jpeg, IIRC: the top-left tag in EXIF, a rotate-90 tag somewhere, and one other. I think the 1st one is the most popular today. The 2nd one was widely used e.g. by ACDSEE for years, especially to achieve lossless rotation (which also required both x and y pixel counts to be a multiple of 64). Adobe stopped supporting one of the others after Photoshop 7 (a long time ago) which was annoying if you had a load of old jpegs.

**Comment by snibgo** [comment]
What version of IM do you use, on what platform?

With IM v7.0.8-64 on Windows 8.1, with JPEG files (and probably other formats) IM reads and writes the exif "orientation" flag. If "-auto-orient" is not used, the value of the flag from the first input is written unchanged to the output.

**Comment by peterh337** [comment]
We have this function to deal with rotation on the conversion to jpeg, but it fails for CR2 portrait modes. It does mirroring (flipping) instead.

function autorotate(Imagick $image)
{
    switch ($image->getImageOrientation()) {
    case Imagick::ORIENTATION_TOPLEFT:
        break;
    case Imagick::ORIENTATION_TOPRIGHT:
        $image->flopImage();
        break;
    case Imagick::ORIENTATION_BOTTOMRIGHT:
        $image->rotateImage(new ImagickPixel('#00000000'), 180);
        break;
    case Imagick::ORIENTATION_BOTTOMLEFT:
        $image->flopImage();
        $image->rotateImage(new ImagickPixel('#00000000'), 180);
        break;
    case Imagick::ORIENTATION_LEFTTOP:
        $image->flopImage();
        $image->rotateImage(new ImagickPixel('#00000000'), -90);
        break;
    case Imagick::ORIENTATION_RIGHTTOP:
        $image->rotateImage(new ImagickPixel('#00000000'), 90);
        break;
    case Imagick::ORIENTATION_RIGHTBOTTOM:
        $image->flopImage();
        $image->rotateImage(new ImagickPixel('#00000000'), 90);
        break;
    case Imagick::ORIENTATION_LEFTBOTTOM:
        $image->rotateImage(new ImagickPixel('#00000000'), -90);
        break;
    default: // Invalid orientation
        break;
    }
    $image->setImageOrientation(Imagick::ORIENTATION_TOPLEFT);
    return $image;
}

**Comment by snibgo** [comment]
> ... but it fails for CR2 portrait modes.

Which mode is that? Eg ORIENTATION_BOTTOMLEFT or whatever? Perhaps your code has a bug. Or the camera has a bug.

**Comment by peterh337** [comment]
We are looking at that very Q. It is two portrait images.
Could be a camera bug but CR2 is Canon and they cannot be wrong ;)

**Comment by peterh337** [comment]
The two images which are getting flipped return these orientation codes

IMG_2481.CR2 Imagick::ORIENTATION_RIGHTTOP
IMG_2482.CR2 Imagick::ORIENTATION_LEFTBOTTOM

I am happy to PM the dropbox link to someone interested. I have four CR2s and four JPGs, all ex-camera (G7X Mk1). Of the four, two are portrait and show the issue, and the other two are landscape and work ok. 

One interesting discovery, perhaps not surprising, is that if the camera is used totally upside down, it produces a file with a normal landscape orientation but the image is upside down. This is not what a phone does; that will always try to deliver an upright image, and again this is obvious when you look at the usage.

**Comment by snibgo** [comment]
Top-right orientation is corrected with a mirroring. I doubt that a camera can make an image that requires miroring. (Unless the photo was taken through a mirror, of course.)

If you send me a link to the problem raw photos, I can take a look.

Your upside-down issue suggests a camera bug or fault.

**Comment by peterh337** [comment]
Sorry; I can't see any way to PM you, here on via your own website. Well not without joining up some other forums.

AIUI the orientation value is the location of top left, so top right implies mirroring L-R (as you say), and left bottom implies mirroring U-D. This is nonsense. There is no config on the camera though. and Canon's own software (Digital Photo Pro 4) handles it correctly.

**Comment by snibgo** [comment]
Ah, no, I can't find a private message button on this forum. Weird.

You can email the address of raw files to XXXX. I will edit this post to remove that address when you've done it.  I don't generally encourage use of my email.

Yes, "orientation top-right" means "move top-right pixel to top-left, by mirroring".

By contrast, "orientation right-top" means "move top-right pixel to top-left by rotation".


>  Canon's own software (Digital Photo Pro 4) handles it correctly.

I'm an old cynic. If a Canon camera has a bug, Canon's own software might work around that bug.

**Comment by peterh337** [comment]
Email sent :)

I agree re Canon, especially as nobody else uses CR2. Unlike DNG... but DNG has some new flavours too which likely break stuff...

Interesting thing about CR2 is that the images look really great, whereas DNGs are mostly washed out and need work in Lightroom. IOW, you have to provide a user interface for enhancing them.

**Comment by peterh337** [comment]
Hello,

Many thanks for your reply.

You can see the two incorrect files at bottom right here

https://airports.euroga.org/record.php?id=119

I posted the PHP function used to do it in the forum.

function autorotate(Imagick $image)
{
switch ($image->getImageOrientation()) {
case Imagick::ORIENTATION_TOPLEFT:
break;
case Imagick::ORIENTATION_TOPRIGHT:
$image->flopImage();
break;
case Imagick::ORIENTATION_BOTTOMRIGHT:
$image->rotateImage(new ImagickPixel('#00000000'), 180);
break;
case Imagick::ORIENTATION_BOTTOMLEFT:
$image->flopImage();
$image->rotateImage(new ImagickPixel('#00000000'), 180);
break;
case Imagick::ORIENTATION_LEFTTOP:
$image->flopImage();
$image->rotateImage(new ImagickPixel('#00000000'), -90);
break;
case Imagick::ORIENTATION_RIGHTTOP:
$image->rotateImage(new ImagickPixel('#00000000'), 90);
break;
case Imagick::ORIENTATION_RIGHTBOTTOM:
$image->flopImage();
$image->rotateImage(new ImagickPixel('#00000000'), 90);
break;
case Imagick::ORIENTATION_LEFTBOTTOM:
$image->rotateImage(new ImagickPixel('#00000000'), -90);
break;
default: // Invalid orientation
break;
}
$image->setImageOrientation(Imagick::ORIENTATION_TOPLEFT);
return $image;
}

FWIW, this has been tested with jpegs from various things
- android phones
- ios phones and ipads
- the G7X camera (it does jpegs and cr2 concurrently)
and all of them work fine.

But they probably use only four of the above eight cases and my guess
is that the cr2 files are invoking two of the other four cases and
there is a mistake there.

Regards,

Peter


>According to "exiftool -args -G0 XXX.cr2", each of those 4 raw files have orientation metadata in two tags: EXIF:Orientation and MakerNotes:AutoRotate.
>
>File 2481 has:
>Orientation=6 ("Rotate 90 CW") AutoRotate=1 ("Rotate 90 CW"). As you say, this is ORIENTATION_RIGHTTOP.
>
>2482 has:
>Orientation=8 ("Rotate 270 CW") AutoRotate=3 ("Rotate 270 CW"). As you say, this is ORIENTATION_LEFTBOTTOM.
>
>Files 2483 and both 2484 have:
>Orientation=1 ("Horizontal (normal)") AutoRotate=0 ("None")
>
>Within each file, the meanings of the tags seem to agree. Most software will ignore MakerNotes.
>
>Using dcraw with "-t 0", it makes a tiff without rotating the pixels. I view these with a home-made viewer that I know ignores orientation metadata. I do this for all 4 raws.
>
>2481 needs 90 CW rotation. The metadata is correct.
>
>2482 needs 270 CW rotation. The metadata is correct.
>
>2483 doesn't need rotating. The metadata is correct.
>
>2484 was taken with the camera inverted (rotated 180 degrees), but the raw metadata says it doesn't need to be rotated. Therefore the raw metadata is wrong. Assuming the file hasn't been edited, the camera is wrong: a bug, or fault, or a design limitation.
>
>I conclude that the camera metadata is okay except where the image needs 180 deg rotation.
>
>For 2481 and 2482, if software is not rotating correctly, then the software has a problem. Is it actually mirroring, or is is rotating twice?

**Comment by peterh337** [comment]
Funny, my reply email ended up here in the forum. Doesn't matter but might be a surprise to some others when they reply to a forum email!

**Comment by snibgo** [comment]
And I thought I had posted the following. Maybe the bug is in my brain...


According to "exiftool -args -G0 XXX.cr2", each of those 4 raw files have orientation metadata in two tags: EXIF:Orientation and MakerNotes:AutoRotate.

File 2481 has:
Orientation=6 ("Rotate 90 CW") AutoRotate=1 ("Rotate 90 CW"). As you say, this is ORIENTATION_RIGHTTOP.

2482 has:
Orientation=8 ("Rotate 270 CW") AutoRotate=3 ("Rotate 270 CW"). As you say, this is ORIENTATION_LEFTBOTTOM.

Files 2483 and both 2484 have:
Orientation=1 ("Horizontal (normal)") AutoRotate=0 ("None")

Within each file, the meanings of the tags seem to agree. Most software will ignore MakerNotes.

Using dcraw with "-t 0", it makes a tiff without rotating the pixels. I view these with a home-made viewer that I know ignores orientation metadata. I do this for all 4 raws.

2481 needs 90 CW rotation. The metadata is correct.

2482 needs 270 CW rotation. The metadata is correct.

2483 doesn't need rotating. The metadata is correct.

2484 was taken with the camera inverted (rotated 180 degrees), but the raw metadata says it doesn't need to be rotated. Therefore the raw metadata is wrong. Assuming the file hasn't been edited, the camera is wrong: a bug, or fault, or a design limitation.

I conclude that the camera metadata is okay except where the image needs 180 deg rotation.

For 2481 and 2482, if software is not rotating correctly, then the software has a problem. Is it actually mirroring, or is is rotating twice? The code you have posted looks okay, but what happens prior to that?

**Comment by peterh337** [comment]
What concerns me is breaking the code for all the other images which currently work.
I have put eight HEIC images in the same dropbox link. Would you be able to see if any of these use the two currently incorrect CR2 case statements?

**Comment by snibgo** [comment]
To do one, in a human-readable format, do this:

```
exiftool -args -G0 IMG_0823.HEIC |grep Orientation
```

The result is:
```
-EXIF:Orientation=Horizontal (normal)
```

To see the result as a number instead, use "-n":
```
exiftool -args -n -G0 IMG_0823.HEIC |grep Orientation

-EXIF:Orientation=1
```

**Comment by peterh337** [comment]
The camera is set to produce both cr2 and jpg at the same time.

2481.jpg shows (in ACDSEE v7) orientation = right upper
and it works fine in the above code.

So that is the same as 2481.cr2 (right top)

So why doesn't this invoke the following case in the above code for the cr2?

case Imagick::ORIENTATION_RIGHTTOP:
$image->rotateImage(new ImagickPixel('#00000000'), 90);
break;

It would appear that Imagemagick may be doing the rotation incorrectly for the cr2.

**Comment by snibgo** [comment]
I suggest you create a testing environment, if you haven't already, then insert tracing commands in your code. Then you can run a file trough your script, and trace the value of getImageOrientation() immediately before the switch. You can also write your $image to a file, so you can see if pixels are where you expect them to be.

I don't use PHP, so can't advise on the actual commands. But one principle of debugging is: when a certain input gives an unexpected output, insert traces to isolate the step that doesn't do what you expect.

**Comment by peterh337** [comment]
Imagemagick is v6.

**Comment by peterh337** [comment]
More data on Imagemagick

imagick module version 3.4.4
imagick classes Imagick, ImagickDraw, ImagickPixel, ImagickPixelIterator, ImagickKernel
Imagick compiled with ImageMagick version ImageMagick 6.9.10-86 Q16 x86_64 2020-01-13 https://imagemagick.org
Imagick using ImageMagick library version ImageMagick 6.9.11-16 Q16 x86_64 2020-06-01 https://imagemagick.org
ImageMagick copyright © 1999-2020 ImageMagick Studio LLC
ImageMagick release date 2020-06-01

**Comment by peterh337** [comment]
Centos 8
PHP 7.4.7 
Imagick compiled with ImageMagick version => ImageMagick 6.9.11-16 Q16 x86_64 2020-06-01 https://imagemagick.org
Imagick using ImageMagick library version => ImageMagick 6.9.11-16 Q16 x86_64 2020-06-01 https://imagemagick.org

relevant packages:
ImageMagick.x86_64                                              1:6.9.11.16-1.el8.remi                                         @remi                  
ImageMagick-devel.x86_64                                        1:6.9.11.16-1.el8.remi                                         @remi                  
ImageMagick-heic.x86_64                                         1:6.9.11.16-1.el8.remi                                         @remi                  
ImageMagick-libs.x86_64                                         1:6.9.11.16-1.el8.remi                                         @remi                  

php74-php-pecl-imagick.x86_64                                   3.4.4-10.el8.remi                                              @remi-safe             

Problem:
https://airports.euroga.org/IMG_2481.CR2 - original file
https://airports.euroga.org/images/265/2.jpeg - this is the result of browser file upload and conversion to jpg.
https://airports.euroga.org/2481_conv_test_no_res_no_rotate.jpeg - same file, uploaded through ssh and converted for a command line PHP script.
EXIF Orientation: 6 (according to both IMagick and exif_read_data)

Another set:
https://airports.euroga.org/IMG_2482.CR2
https://airports.euroga.org/images/265/3.jpeg
https://airports.euroga.org/2482_conv_test_no_res_no_rotate.jpeg
EXIF Orientation: 8

I've reduced script to minimum:
$filepath = 'IMG_2482.CR2';
$im = new Imagick($filepath);
$im->writeImage('2482_conv_test_no_res_no_rotate.jpeg');
and same goes for the uploaded version, just it comes through $_FILES

As you can see, without any manual rotation or any other alterations, the result is different for both files.

**Comment by snibgo** [comment]
The file IMG_2481.CR2 (5472x3648 pixels) has orientation "rotate 90 CW".

The file 2.jpeg (5472x3648) has orientation "undefined". You say "this is the result of browser file upload and conversion to jpg." What software read the CR2 file? That software has dropped the orientation metadata. The image pixels have not been rotated, so it looks wrong.

The file 2481_conv_test_no_res_no_rotate.jpeg (3693x5536) also has orientation "undefined" (according to IM and exiftool), but the pixels have been rotated, so it looks okay.

You say 2481_conv_test_no_res_no_rotate.jpeg is "EXIF Orientation: 6 (according to both IMagick and exif_read_data)". That is different to the result I get from ImageMagick and exiftool.

**Comment by peterh337** [comment]
It has been confirmed that 2481.cr2 takes this parth through the case statement

case Imagick::ORIENTATION_RIGHTTOP:
$image->rotateImage(new ImagickPixel('#00000000'), 90);
break;

and the jpeg produced is flipped, not rotated as it should be. 

The 2481.jpg file, generated concurrently by the G7X camera, takes the same path through the code, and produces a jpeg with the correct orientation.

This suggests that IM performs the wrong operation on a cr2 file. It doesn't surprise me; you will hardly have the whole world using IM with CR2 files :)

**Comment by snibgo** [comment]
> This suggests that IM performs the wrong operation on a cr2 file.

No. Once ImageMagick (via IMagick) has read the image from the file, the file format is no longer relevant. "rotateImage()" will do the same action, whatever the file format was.

**Comment by peterh337** [comment]
It does look like a bug in IM on the CR2 handling. The following version of the above function, used only for CR2 processing, delivers correctly orientated files. Only the two portrait cases were treated differently.

function autorotate_cr2(Imagick $image)

{
    switch ($image->getImageOrientation()) {
    case Imagick::ORIENTATION_TOPLEFT:
        log_write('Imagick::ORIENTATION_TOPLEFT');
        break;
    case Imagick::ORIENTATION_TOPRIGHT:
        log_write('Imagick::ORIENTATION_TOPRIGHT');
        $image->flopImage();
        break;
    case Imagick::ORIENTATION_BOTTOMRIGHT:
        log_write('Imagick::ORIENTATION_BOTTOMRIGHT');
        $image->rotateImage(new ImagickPixel('#00000000'), 180);
        break;
    case Imagick::ORIENTATION_BOTTOMLEFT:
        log_write('Imagick::ORIENTATION_BOTTOMLEFT');
        $image->flopImage();
        $image->rotateImage(new ImagickPixel('#00000000'), 180);
        break;
    case Imagick::ORIENTATION_LEFTTOP:
        log_write('Imagick::ORIENTATION_LEFTTOP');
        $image->flopImage();
        $image->rotateImage(new ImagickPixel('#00000000'), -90);
        break;
    case Imagick::ORIENTATION_RIGHTTOP:
        log_write('Imagick::ORIENTATION_RIGHTTOP');
        $image->flopImage();
        $image->rotateImage(new ImagickPixel('#00000000'), 90);
        break;
    case Imagick::ORIENTATION_RIGHTBOTTOM:
        log_write('Imagick::ORIENTATION_RIGHTBOTTOM');
        $image->flopImage();
        $image->rotateImage(new ImagickPixel('#00000000'), 90);
        break;
    case Imagick::ORIENTATION_LEFTBOTTOM:
        log_write('Imagick::ORIENTATION_LEFTBOTTOM');
        $image->flopImage();
        $image->rotateImage(new ImagickPixel('#00000000'), 90);
        break;
    default: // Invalid orientation
        break;
    }
    $image->setImageOrientation(Imagick::ORIENTATION_TOPLEFT);
    return $image;
}
